### PR TITLE
CDAP-13261 Support reading metadata in program (in-prem only) Part 1

### PR DESCRIPTION
**Pull request by rohitsinha54**
## Summary
- Adds support for reading metadata in CDAP programs
- In 5.0 read is only supported in in-prem mode and not cloud mode.
- For the in-prem mode the implementation of MetadataReaderContext uses MetadataAdmin to get the metadata directly from the store.
- Added a unit test where a `service` reads the metadata

## Misc
- Part 2 will add support for emitting metadata to TMS (broken down to keep pr small)
- Part 3 will add more extensive unit tests which will have unit tests for all the programs reading and writing metadata (broken down to keep pr small)
- Currently, have tested modified PurchaseHistoryBuilder `mapreduce` job being able to access metadata in standalone and in-prem (single node).

## Tasks:
- [x] Build: https://builds.cask.co/browse/CDAP-DUT6497-2
- [x] Integration Tests:
- [x] Test on multinode cluster
- [ ] Test behavior inside spark's closures

### CODE REVIEW

**Comment by rohitsinha54** [comment]
Closing PR will reopen with changes to use MetadataClient rather than Admin.

File: cdap-api-spark-base/src/main/java/co/cask/cdap/api/spark/JavaSparkExecutionContextBase.java [hunk 1]
```diff
@@ -110,6 +112,14 @@
    */
   public abstract MessagingContext getMessagingContext();
 
+  /**
+   * Returns a {@link MetadataReaderContext} which can be used to read metadata. Currently the returned instance can
+   * only be used in the Spark driver process.
+   *
+   * @return A {@link MetadataReaderContext}
+   */
+  public abstract MetadataReaderContext getMetadataReaderContext();
```
Comment (anew) [inline cdap-api-spark-base/src/main/java/co/cask/cdap/api/spark/JavaSparkExecutionContextBase.java]: is this a user-facing API? Feels like it should just be called getMatadataReader(). Actually, this context implements the MetadataReaderContext. So why is there a need for a public getMetadataReaderContext?

File: cdap-api/src/main/java/co/cask/cdap/api/mapreduce/MapReduceTaskContext.java [hunk 2]
```diff
@@ -41,7 +42,7 @@
  */
 @Beta
 public interface MapReduceTaskContext<KEYOUT, VALUEOUT> extends SchedulableProgramContext, RuntimeContext,
-  DatasetContext, ServiceDiscoverer, PluginContext, TaskLocalizationContext, SecureStore {
+  DatasetContext, ServiceDiscoverer, PluginContext, TaskLocalizationContext, SecureStore, MetadataReaderContext {
```
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/mapreduce/MapReduceTaskContext.java]: should this be in the task context? That is, every task can read meta data? That can cause a huge load of requests... I would like to encourage developers to call it in the driver and serialize it to all tasks.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java [hunk 3]
```diff
@@ -161,7 +166,8 @@ protected AbstractContext(Program program, ProgramOptions programOptions, CConfi
                             @Nullable MetricsCollectionService metricsService, Map<String, String> metricsTags,
                             SecureStore secureStore, SecureStoreManager secureStoreManager,
                             MessagingService messagingService,
-                            @Nullable PluginInstantiator pluginInstantiator) {
+                            @Nullable PluginInstantiator pluginInstantiator,
+                            MetadataReaderContext metadataReaderContext) {
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java]: why is this called a "context"? Isn't it simply a meta data reader?

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/InPremMetadataReader.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Implementation for {@link MetadataReaderContext} which used {@link MetadataAdmin} to read metadata.
+ * This implementation should only be used while running in-prem mode where the {@link MetadataStore} is accessible to
+ * the program container.
+ * Note: This implementation should not be used in cloud mode.
+ */
+public class InPremMetadataReader implements MetadataReaderContext {
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/InPremMetadataReader.java]: not a very good name for the class. It should describe what the class is, not where it is used (that may change later and then it would be confusing). I would call this a RemoteMetadataReader, or a MetadataServiceClient.

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/InPremMetadataReader.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+  @Inject
+  public InPremMetadataReader(MetadataAdmin metadataAdmin) {
+    this.metadataAdmin = metadataAdmin;
+  }
+
+  @Override
+  public Map<MetadataScope, Metadata> getMetadata(MetadataEntity metadataEntity) {
+    Map<MetadataScope, Metadata> scopeMetadata = new HashMap<>();
+    Set<MetadataRecordV2> metadata = metadataAdmin.getMetadata(metadataEntity);
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/InPremMetadataReader.java]: I don't think this will work. This will be injected into program containers, which run as the (imperonated) user. But only the CDAP system user can read the MetadataDataset (which is needed to instantiate a MetadataAdmin). You will have to make an HTTP call to the metadata service.

File: cdap-spark-core-base/src/main/scala/co/cask/cdap/app/runtime/spark/DefaultJavaSparkExecutionContext.scala [hunk 6]
```diff
@@ -75,6 +76,8 @@ class DefaultJavaSparkExecutionContext(sec: SparkExecutionContext) extends JavaS
 
   override def getMessagingContext: MessagingContext = sec.getMessagingContext
 
+  override def getMetadataReaderContext: MetadataReaderContext = sec.getMetadataReaderContext
```
Comment (anew) [inline cdap-spark-core-base/src/main/scala/co/cask/cdap/app/runtime/spark/DefaultJavaSparkExecutionContext.scala]: why is this needed?

**Review by anew — COMMENTED** [review]
please see comments, major concerns are:
- MR tasks should not read meta data
- authorization will fail if programs directly instantiate MetadataAdmin
- naming (but not that important unless the user facing API exposes the name)

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java (hunk 3 above)
Comment (rohitsinha54) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java]: Done.

File: cdap-api/src/main/java/co/cask/cdap/api/mapreduce/MapReduceTaskContext.java (hunk 2 above)
Comment (rohitsinha54) [reply cdap-api/src/main/java/co/cask/cdap/api/mapreduce/MapReduceTaskContext.java]: Agreed, removed.

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/InPremMetadataReader.java (hunk 4 above)
Comment (rohitsinha54) [reply cdap-app-fabric/src/main/java/co/cask/cdap/metadata/InPremMetadataReader.java]: Done.

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/InPremMetadataReader.java (hunk 5 above)
Comment (rohitsinha54) [reply cdap-app-fabric/src/main/java/co/cask/cdap/metadata/InPremMetadataReader.java]: Ya right. Changed it.

**Review by rohitsinha54 — COMMENTED** [review]
@anew Addressed comments. Please have a look.

File: cdap-spark-core-base/src/main/scala/co/cask/cdap/app/runtime/spark/DefaultJavaSparkExecutionContext.scala (hunk 6 above)
Comment (rohitsinha54) [reply cdap-spark-core-base/src/main/scala/co/cask/cdap/app/runtime/spark/DefaultJavaSparkExecutionContext.scala]: I sticked with the pattern of exposing the interfaces being implemented. I think the idea is that we want to get these so that we can pass it to the transformations (given the object itself is serializable) although since we have already decided that we will not emit metadata at record level we don't need it as such. 

Removed.

File: cdap-api-spark-base/src/main/java/co/cask/cdap/api/spark/JavaSparkExecutionContextBase.java (hunk 1 above)
Comment (rohitsinha54) [reply cdap-api-spark-base/src/main/java/co/cask/cdap/api/spark/JavaSparkExecutionContextBase.java]: Removed. See comment on another similar one.

**Comment by rohitsinha54** [comment]
Build: https://builds.cask.co/browse/CDAP-DUT6497-6

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataException.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+ * the License.
+ */
+
+package co.cask.cdap.api.metadata;
+
+
+/**
+ * Exception which is thrown when a metadata operation fails
+ */
+public class MetadataException extends RuntimeException {
```
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataException.java]: why is it a RuntimeException?

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataReaderWriterModules.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+      }
+    };
+  }
+
+  @Override
+  public Module getDistributedModules() {
+    return new AbstractModule() {
+      @Override
+      protected void configure() {
+        // CDAP-13610 We should be binding MetadataAdmin which the AbstractMetadataAdmin implements.
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataReaderWriterModules.java:57]: I am not sure I understand this comment... why do the in-memory and local modules not bind this, and where is this used? Do we inject a AbstractMetadataClient anywhere?

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/RemoteMetadataReader.java [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    LOG.trace("Returning metadata record {} for {}", scopeMetadata, metadataEntity);
+    return scopeMetadata;
+  }
+
+  @Override
+  public Metadata getMetadata(MetadataScope scope, MetadataEntity metadataEntity) {
+    final Metadata[] metadata = new Metadata[1];
+    try {
+      metadataClient.getMetadata(metadataEntity, scope).forEach(record -> {
+        if (record.getScope() == scope) {
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/RemoteMetadataReader.java]: can it ever return a record with a different scope? if not, then these two method can share the implementation

**Review by anew — COMMENTED** [review]
only a few comments

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/RemoteMetadataReader.java (hunk 9 above)
Comment (rohitsinha54) [reply cdap-app-fabric/src/main/java/co/cask/cdap/metadata/RemoteMetadataReader.java]: Done.

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataReaderWriterModules.java (hunk 8 above)
Comment (rohitsinha54) [reply cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataReaderWriterModules.java:57]: We need two implementation of the MetadataClient (rest based and another remote which can do discovery). Currently we don't have any interface for metadata client its just one gigantic class. All the logic reside in AbstractMetadataClient and there is a MetadataClient which extends it and uses restClient serve the request. The RemoteMetadataClient which is introduced in this PR uses a remoteClient and service discovery to serve the request.

AbstractMetadataClient is injected in RemoteMetadataReader to support the calls made from programatic apis. In case of local and standalone mode the DefaultMetadataReader is used. Since neither of these mode is distributed and doesn't have impersonation DefaultMetadataReader directly reads the store through MetadataAdmin rather than going through a client.

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataException.java (hunk 7 above)
Comment (rohitsinha54) [reply cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataException.java]: Rethinking client might be able to handle it and recover in some cases. Changed it Exception.

**Review by anew — APPROVED** [review]
